README.md

```diff
@@ -1,11 +1,74 @@
+perform custom validation or mutation, and modify the response as needed, thus signaling if validation succeeded
+or failed. When the Webhook is done, it returns control to Wrangler which then handles the sending of the verdict
+of the admission controller.
+
+```Admit(webhookResponse *Response, webhookRequest *Request) error```
+
+Note that each object's Admit method returns an error.
+This error is not to signify that an object violates some validation rules, but rather to signal that
+the validation process itself has encountered an error. If an admission controller has deemed a resource
+invalid for persistence, it means it itself has succeeded.
+
+Any admission controller, as an app, consists of two main things:
+1. The configuration which describes the resources and actions for which the webhook is active. This
+   configuration also references the Kubernetes service which directs traffic to the actual web-server
+   (this Webhook project) that does the work. The configuration exists as ValidatingWebhookConfiguration
+   and MutatingWebhookConfiguration global resources.
+2. The actual web-server with all the validation and mutation logic found in each object's Admit method.
+
+Both the ValidatingWebhookConfiguration and MutatingWebhookConfiguration are fully created at
+webhook startup, not beforehand. Usually these key admission controller resources are deployed as manifests along with those
+for the web-server deployment and service, but Rancher is different here.
+
+All objects with custom validation logic exist in the `pkg/resources` package. For example, there is a
+`globalrole` package that has a type with an Admit method. Every time the local cluster's kube-api server receives a
+request to create or update a Global Role (a Rancher resource), the Webhook, being an instance of an
+admission controller, intercepts the request. The Admit method for Global Roles runs.
+It inspects the Global Role object and leads it through a series of custom checks. One of them
+ensures that each rule of the Global Role has at least one verb. If it does not, then the Webhook
+changes the response value by setting its `Allowed` field to false, meaning validation has failed. It then returns.
+
+This logic is the main part of object inspection and admission control, the rest is wrapped by
+Wrangler.
+
+The resources that the Webhook validates or mutates are declared in the `pkg/resources/validation`
+and `pkg/resources/mutation`, respectively.
+
+The `pkg/server` package is the main setup package of the Webhook server itself as well as all the low-level
+components defined in Wrangler. The package defines the rules for resources and actions for which the Webhook will
+be active. These are later brought to life as cluster-wide Kubernetes resources
+(ValidatingWebhookConfiguration and MutatingWebhookConfiguration).
+It also binds resources with their respective admission handlers - this is done for the Webhook
+router, which is also defined in Wrangler.
```

Comment:
I'd put this section up higher. One of my first questions was: is there a list of all objects we validate/mutate.